presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -194,19 +198,52 @@ private SubPlan analyzeGroupedExecution(Session session, SubPlan subPlan)
                     && isDynamicScheduleForGroupedExecution(session);
             BucketNodeMap bucketNodeMap = nodePartitioningManager.getBucketNodeMap(session, fragment.getPartitioning(), preferDynamic);
             if (bucketNodeMap.isDynamic()) {
-                fragment = fragment.withDynamicLifespanScheduleGroupedExecution(properties.getCapableTableScanNodes());
+                // We currently only support recoverable grouped execution if the following statements hold true:
+                //   - Current session enables recoverable grouped execution
+                //   - Parent sub plan contains TableFinishNode
+                //   - Current sub plan's root is TableWriterNode
+                //   - Input connectors supports split source rewind
+                //   - Output connectors supports lifespan commit
+                //   - Bucket node map uses dynamic scheduling
+                boolean recoverable = isRecoverableGroupedExecutionEnabled(session) &&
+                        parentContainsTableFinish &&
+                        fragment.getRoot() instanceof TableWriterNode &&
+                        properties.isPreferRecoverable();
+                if (recoverable) {
+                    fragment = fragment.withRecoverableGroupedExecution(properties.getCapableTableScanNodes());
+                }
+                else {
+                    fragment = fragment.withDynamicLifespanScheduleGroupedExecution(properties.getCapableTableScanNodes());
+                }
             }
             else {
                 fragment = fragment.withFixedLifespanScheduleGroupedExecution(properties.getCapableTableScanNodes());
             }
         }
         ImmutableList.Builder<SubPlan> result = ImmutableList.builder();
+        boolean containsTableFinishNode = containsTableFinishNode(fragment);
         for (SubPlan child : subPlan.getChildren()) {
-            result.add(analyzeGroupedExecution(session, child));
+            result.add(analyzeGroupedExecution(session, child, containsTableFinishNode));
         }
         return new SubPlan(fragment, result.build());
     }
 
+    private static boolean containsTableFinishNode(PlanFragment planFragment)
+    {
+        return containsTableFinishNode(planFragment.getRoot());
```

Comment:
per offline discussion, let's just do a "pattern matching" of "OutputNode - TableFinishNode" 😃